core/services/ocr2/plugins/ocr2keeper/evm/registry.go

```diff
@@ -0,0 +1,771 @@
+		}
+		return nil
+	})
+}
+
+func (r *EvmRegistry) Ready() error {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	if r.runState == 1 {
+		return nil
+	}
+	return r.sync.Ready()
+}
+
+func (r *EvmRegistry) Healthy() error {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	if r.runState > 1 {
+		return fmt.Errorf("failed run state: %w", r.runError)
+	}
+	return r.sync.Healthy()
+}
+
+func (r *EvmRegistry) initialize() error {
+	startupCtx, cancel := context.WithTimeout(r.ctx, 30*time.Second)
+	defer cancel()
+
+	idMap := make(map[int64]activeUpkeep)
+
+	// get active upkeep ids from contract
+	ids, err := r.getLatestIDsFromContract(startupCtx)
+	if err != nil {
+		return fmt.Errorf("failed to get ids from contract: %s", err)
+	}
+
+	var offset int
+	for offset < len(ids) {
+		batch := 100
+		if len(ids)-offset < batch {
+			batch = len(ids) - offset
+		}
+
+		actives, err := r.getUpkeepConfigs(startupCtx, ids[offset:offset+batch])
+		if err != nil {
+			return fmt.Errorf("failed to get configs for id batch (length '%d'): %s", batch, err)
+		}
+
+		for _, active := range actives {
+			idMap[active.ID.Int64()] = active
+		}
+
+		offset += batch
+	}
+
+	r.mu.Lock()
+	r.active = idMap
+	r.mu.Unlock()
+
+	return nil
+}
+
+func (r *EvmRegistry) pollLogs() error {
+	var start int64
+	var end int64
+	var err error
+
+	if end, err = r.poller.LatestBlock(); err != nil {
+		return fmt.Errorf("%w: %s", ErrHeadNotAvailable, err)
+	}
+
+	r.mu.Lock()
+	start = r.lastPollBlock
+	r.lastPollBlock = end
+	r.mu.Unlock()
+
+	// if start and end are the same, no polling needs to be done
+	if start == 0 || start == end {
+		return nil
+	}
+
+	{
+		var logs []logpoller.Log
+
+		if logs, err = r.poller.LogsWithSigs(
+			start,
+			end,
+			upkeepStateEvents,
+			r.addr,
+			pg.WithParentCtx(r.ctx),
+		); err != nil {
+			return fmt.Errorf("%w: %s", ErrLogReadFailure, err)
+		}
+
+		for _, log := range logs {
+			r.chLog <- log
+		}
+	}
+
+	r.mu.Lock()
+	r.lastPollBlock = end
+	r.mu.Unlock()
+
+	return nil
+}
+
+func (r *EvmRegistry) registerEvents(addr common.Address) error {
+	// Add log filters for the log poller so that it can poll and find the logs that
+	// we need
+	filterID, err := r.poller.RegisterFilter(logpoller.Filter{
+		EventSigs: append(upkeepStateEvents, upkeepActiveEvents...),
+		Addresses: []common.Address{addr},
+	})
+	if err != nil {
+		r.mu.Lock()
+		r.filterID = filterID
+		r.mu.Unlock()
+	}
+	return err
+}
+
+func (r *EvmRegistry) processUpkeepStateLog(l logpoller.Log) error {
+
+	rawLog := l.ToGethLog()
+	abilog, err := r.registry.ParseLog(rawLog)
+	if err != nil {
+		return err
+	}
+
+	switch l := abilog.(type) {
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepMigrated: // removes upkeep id and detail from registry
+		r.removeFromActive(l.Id)
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepCanceled:
+		r.removeFromActive(l.Id)
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepPaused:
+		r.removeFromActive(l.Id)
```

Comment:
to potentially simplify we can remove this logic. Deletion of upkeeps will automatically happen on refresh interval so this won't have any visible user effect. (apart from very minor performance gain)

Reply:
Removing this will help in making this function log order independent, and we won't have issues processing logs twice making it easy to tackle re-orgs